server/src/main/java/org/jboss/as/server/deployment/module/ManifestClassPathProcessor.java

```diff
@@ -141,11 +141,17 @@ public synchronized void deploy(final DeploymentPhaseContext phaseContext) throw
             }
         }
 
+        Set<VirtualFile> processedClassPathFiles = new HashSet<>();
+        Attachable lastTarget = null;
         //additional resource roots may be added as
         while (!resourceRoots.isEmpty()) {
             final RootEntry entry = resourceRoots.pop();
             final ResourceRoot resourceRoot = entry.resourceRoot;
             final Attachable target = entry.target;
+            if (target != lastTarget) {
+                processedClassPathFiles.clear();
+                lastTarget = target;
+            }
```

Comment:
Please add a comment explaining this logic, as this class is already really complex and it's not intuitive why this is done instead of just creating processedClassPathFiles inside the loop. Reviewing this carefully I get that the idea is the same Attachable can be attached to multiple resource roots, but it's not clear why it's only appropriate to check 'lastTarget'.

Perhaps this should use a Map<Attachable, Set<VirtualFile>>> for this tracking?

Reply:
Yes, you are right, Map<Attachable, Set>> is correct.

target != lastTarget worked for me by chance in my deployment, and in some other case it could have failed to achieve anything useful.